### DISCUSSION TITLE
Health Tasks workflow

### DISCUSSION THREAD

**Discussion by dubdabasoduba** [opensrp/fhircore#1557] [Ideas]
Defining the flow of Tasks through the different statuses
---
**Tasks status**
- **Requested** -- This status is given to a Task when the `executionPeriod.start` date is in the future. 
- **Ready** -- This status is given to a Task when the `executionPeriod.start` date is <= to today and the `executionPeriod.end` date >= today
- **In Progress** -- This status is given to a Task when the QuestionnaireResponse it acts on is marked as `draft`
- **Completed** -- This status is given to a Task when the QuestionnaireResponse it acts on is marked as `completed`
- **Overdue** -- Is status is not assigned to the Task resource. This status is inferred from the following logic. A **Ready** / **In Progress** task that the `executionPeriod.end` date < today
- **Expired** -- This status will be represented by the `cancelled` status on the Task resource. This is a Task that has been Overdue for a while and therefore it can't be performed anymore. 

**Tasks Color Coding**
- The following designs and colour codes should be used for the different Task statuses. 
    - **Requested** -- This should have a light gray background with Gray font & + sign. Here is a sample
          ![Requested Task](https://user-images.githubusercontent.com/4817531/185799592-8e003d42-1ab1-4167-8767-6df95b7d9907.png)
    - **Ready** -- This should have a light blue background with Blue font & + sign. Here is a sample
          ![Ready Task](https://user-images.githubusercontent.com/4817531/185799602-33e74d14-5dd9-45e0-854a-cb952e1a2d21.png)
    - **In-Progress** -- This should have a light orange background with an Orange font & loading sign. Here is a sample
          ![In-Progress Task](https://user-images.githubusercontent.com/4817531/185799611-325c60b4-0356-49b4-a603-db76b48af835.png) 
    - **Completed** -- This should have a light grey background with Gray font & a tick sign. Here is a sample
          ![Completed Task](https://user-images.githubusercontent.com/4817531/185799661-9e013b6d-e734-4704-b475-e0d925c17ed6.png)
   - **Overdue** -- This should have a light red background with Red font & + sign. Here is a sample
        ![Overdue Task](https://user-images.githubusercontent.com/4817531/185799905-26ba2363-948a-4035-9ded-164aa9102b9e.png)
   - **Expired** -- Design TBD cc: @HenryRae 

**Current Workflow.**
- We currently, have a process to generate Tasks & CarePlans via PlanDefinitions & StructureMap Extraction. 
- We have the Tasks linked to Questionnaires as the actions to be performed during the tasks. 
- We currently generate all Tasks for a certain CarePlan all at once.  Then enable the actions based on the Tasks `executionPeriod.end` date. 
- Currently, all the Task linked Questionnaires that required PlanDefinition extraction don't work as required.  This is because the Questionnaires aren't linked to the PlanDefinitions to define any extra schedules required.
    - Examples of these extra schedules can be the _referral schedules_. 
- We also don't have a way to mark QuestionnaireResponses as draft 
- We currently, don't allow editing on complete tasks.  

**Proposed  workflow**

**Tasks generation changes**
- [ ] Generate Tasks on demand instead of generating all the Tasks at once. This means a new task is generated when the current one is `completed` #1696 
     - We however want to make this configurable since some items would want the Tasks generated all at once. e.g If we choose to visualize the child immunization schedules using tasks then we would all the tasks for the schedule to be generated at once. 
     - We will stop the generation of the tasks once the `executionPeriod.end` is >= to the CarePlan `period.end` date.
     - We will always generate the first on the day of CarePlan creation/generation. 
- [ ] Allow the completion of `Overdue` Tasks. This means these Tasks will be clickable and the users could fill up the Questionnaires tied to them. #1697 
     - We will need to update the Task `executionPeriod.end` date to the date the Questionnaire is completed. 
     - We would then need to generate the next Task on the schedule i.e if we had a 3-month schedule with 3 tasks each to be completed each month. An overdue month 1 task completed in month 2 will mean we generate the month 3 task instead of the month 2 task. 
 - [ ] Add the functionality to mark Tasks as **Expired** #1703 
     - Write a service to identify the Tasks that are way past the overdue dates. The service will then update the status to `cancelled`. 
     - We will use the `restriction.period` to hold the allowed lifespan of the Tasks
         - **Start Date** -- This will be equal to the `executionPeriod.start`. It's the DateTime the task is supposed to start. 
         - **End Date** -- This is the last date that the Tasks can be completed. 
 - [ ] We will need to reference the PlanDefitions the Questionnaires on Tasks need to extra any extra schedules based on the QuestionnaireResponses values. #1699 
      - We can use the `Task.input.valueReference` property on the Task. This accepts different kinds of data types one of them being a Resource Reference. 
      -  https://www.hl7.org/fhir/task-definitions.html#Task.input.valueReference
      - We will then update FHIRCore to check whether the Tasks have a PlanDefinition tied to them and then trigger extraction just like the program enrolment Questionnaires right now. E.g how we extract the ANC and PNC schedules. 
- [ ] Extraction of any kind can only be triggered if the QuestionnaireResponse `QuestionnaireResponse.status` property is set to `Complete`  #1700 
- [ ] **NB: -** The old way of generating all tasks might still be needed for some use cases e.g child immunization-related tasks. 

**Tasks state machine changes**
- [ ] Write a service to identify the `Overdue` Tasks. This service can be run every time the Tasks are loaded on each page. 
    - We can have it run periodically but it might not make sense since we are not changing the status on the Task Resource.
    - NB: - We also need to review how this will affect measure-reports since we might have indicators around the number of expired/overdue Tasks
- [ ] Write a service to identify the **Expired** (status `cancelled`) tasks. This service can be run periodically. 
   - The service will then update the Tasks `status` to `cancelled`.  
- [ ] Write a service to identify the `Ready` Tasks. This service can run in 2 different ways #1701 
   - It can run periodically. The Proposal is to have it run after every 1 hour. This time should however be configurable. Some projects might want more/less frequent.  
   - The service can also be run every time a patient profile is open (not recommended). This however has some flaws since the Tasks will not be updated unless the user explicitly opens the subject's profile.  
   - The service should do the following: - 
      -  Find all the Tasks with a `executionPeriod.start` date == to today then change the `Task.status` property to `ready` 
- [ ] Write functionality to check the QuestionnaireResponse `QuestionnaireResponse.status` property to identify whether the QuestionnaireResponse is `In Progress`. The functionality then updates the Task `Task.status` property to `In Progress`. #1702 
- [ ] Update the current functionality that marks Tasks as complete to also check the QuestionnaireResponse `QuestionnaireResponse.status` property to make sure it's set to `Completed` before marking a Task as complete. #1631 

**Questionnaire Response Changes -- SDK Related #1631 
- [ ] Add the ability to mark a QuestionnaireResponse as draft by setting the QuestionnaireResponse `QuestionnaireResponse.status` property to `In Progress`. 
     - Going back or clicking Exit on the Questionnaire page will show a dialog with 2 options.  
          - **Save Partially** -- Allows the user to save the questionnaire partially by setting the QuestionnaireResponse `QuestionnaireResponse.status` property to `in progress`, then redirects to the screen before the questionnaire render. 
          - **Discard** -- This discards all the inputs on the form and closes the Questionnaire render page, then redirects to the screen before the Questionnaire.
          - **Cancel** -- @HenryRae Do we need a cancel button?

**Comment by ellykits** [comment]
What's the corresponding FHIR task status for overdue?

**Reply by dubdabasoduba** [reply]
@ellykits the Overdue task status does not have a corresponding FHIR state. This will be a calculation checking the `executionPeriod.end` and the current status e.g `in progress` & `ready`

**Comment by ndegwamartin** [comment]
There's an interesting library here from Tinder we can check out https://github.com/Tinder/StateMachine

**Comment by pld** [comment]
Thanks for the write-up, based on the description, this sounds broader, more about Task Workflow / Management than the state machine specifically, because the state machine is only one sub-component of the different interactions you are describing, like scheduling and loading of specific questionnaires seem key to this but outside the scope of what is normally a state machine.

With $apply getting closer (maybe it won't be ready in time) we should consider if we can use it for the task generation portions, like rerunning it after data changes to generate the updated tasks. Ignoring the question of whether they're embedded in a CarePlan or not, does $apply generate all or only currently applicable tasks?

**Reply by dubdabasoduba** [reply]
@pld Not very sure whether $apply currently generates the Tasks.  I will review it and revert. @f-odhiambo any ideas?

**Reply by pld** [reply]
Got another update on this today, it is WIP, https://github.com/google/android-fhir/pull/1603

**Comment by dubdabasoduba** [comment]
@pld @f-odhiambo I will go ahead and create the SDK-related issue for it to be worked on.

**Comment by AnnieMungai** [comment]
@dubdabasoduba how do we handle expired tasks - these are the tasks that became overdue but were not completed. If they are not completed within a specific period of time, then they become irrelevant to the user thus are termed as expired. They are no longer seen on the client profile. Example is a referral task that was not followed up after when it was supposed to be followed up.

**Reply by pld** [reply]
This is for us to decide together now, there's no preset way these *need* to be handled. @AnnieMungai what do you think would be best for the user?

It sounds like there's some date, after the date they are due, when a task should be move to "expired/failed" and closed. The complication is that you mention there are "some" tasks like this, which sounds like not all, do we have to manage when happens to an overdue task, and how long after overdue it is marked as failed on a per-task basis? We can have a format for including this metadata in the Task (and in the PlanDefinition that generates the tasks)

**Reply by ageryck** [reply]
If we can brainstorm on all outlier scenarios then decide on a workflow that would fit them.
- [ ] Expiring tasks as mentioned above
- [ ] Cancelled tasks - no longer required
- [ ] sequential tasks - irrespective of how late the current task is completed the next task will still need to happen 
- [ ] non-sequential - example described by Ben above

**Reply by pld** [reply]
To make sure I'm on the same page, here's a visual aid 

```
     task start      task due       task expired
time >----------------->--------------------|
```

after a task is due it is still on the task list and can be completed, after it is expired it can no longer be completed and is no longer on the task list.

**Reply by AnnieMungai** [reply]
@pld @ageryck to follow-up on this discussion, and after looking at som of the use cases; here is some clarification for the expired tasks. cc @dubdabasoduba 
 ```
    task start   task due     task overdue   task expired
time >----------------->----------->--------------------------|
```

Expired tasks can be:
- Expired after a certain period and we do not care about them - so are closed. an example would be referrals. The user will not need to see the expired task, but we may need to capture these in the reports—for example number of expired referrals. 
- Expired but we still care as the user can go back and fill in the data in the expired task - for example - immunizations (if we use scheduling and tasking for immunizations).

**Reply by dubdabasoduba** [reply]
@pld the idea is to have a way to define the due date and the expiry date on the Tasks. The current proposal is to use the following keys on the schema. 

Task start date -- `executionPeriod.start`
Task due date -- `executionPeriod.end`
Task start date -- `restriction.period.start`(same as the `executionPeriod.start`)
Task expiry date -- `restriction.period.end` 

We will write 2 services on the app. 
1. A service to identify Overdue tasks. 
2. A service to identify Expired tasks and mark them as `status == cancelled`

**Reply by dubdabasoduba** [reply]
Sequential tasks will be handled by the on-demand task generation.  i.e the plan definition will be written to check that the prerequisite tasks have been completed.

**Reply by pld** [reply]
cool, followups

1) why both with setting restriction.period.start, is that required? if it's the same as executionPeriod.start I don't see why it's needed

2) On the services, is the importance of identify overdue tasks so that we can officially change their state, and then through that display them correctly in the UI?

3) NB that we need to make firm decisions on how the tasks services are run, the description above lays out the options, but before implementation we need to do the deciding

**Reply by dubdabasoduba** [reply]
@pld 

> why both with setting restriction.period.start, is that required? if it's the same as executionPeriod.start I don't see why it's needed

This is one of those things I added for data completeness purposes. It felt weird to have the period with just the end date.

> On the services, is the importance of identify overdue tasks so that we can officially change their state, and then through that display them correctly in the UI?

Yes, that the use case for the services. 

> NB that we need to make firm decisions on how the tasks services are run, the description above lays out the options, but before implementation we need to do the deciding

Agreed,

**Reply by HenryRae** [reply]
> Generate Tasks on demand instead of generating all the Tasks at once. This means a new task is generated when the current one is completed

@dubdabasoduba how can we handle showing upcoming tasks for instance if Health worker has to inform client when the next visit will happen for proper planning...
Are we still not able to show upcoming tasks in the questionares?
cc: @rowo @AnnieMungai 



>  Allow the completion of Overdue Tasks. This means these Tasks will be clickable and the users could fill up the Questionnaires tied to them.
> We would then need to generate the next Task on the schedule i.e if we had a 3-month schedule with 3 tasks each to be completed each month. An overdue month 1 task completed in month 2 will mean we generate the month 3 task instead of the month 2 task.

@dubdabasoduba what happens incase you have 3 tasks in 1 month. Does the overdue tasks overshadow the others? don't the 2 other new tasks get generated?



> Add the ability to mark a QuestionnaireResponse as draft by setting the QuestionnaireResponse QuestionnaireResponse.status property to In Progress.
> Going back or clicking Exit on the Questionnaire page will show a dialog with 2 options.
> Cancel -- @HenryRae Do we need a cancel button?

@dubdabasoduba yes we do need cancel button to discard the dialog.

**Reply by samkanga** [reply]
@pld @dubdabasoduba I would like to add some perspectives on the tasks discussion; while an expired task need not be shown on the task list, on some implementations such as Immunisation, we  have been showing the expired tasks on the child's profile. This is usually for the care giver to have a glance at what vaccinations have been administered and those that were never administered(expired). This information is helpful in the longitudinal care of the client.

**Reply by pld** [reply]
Thanks, that's useful. I think there are two issues here we can separate, 1) is the tasking of vax, 2) is the schedule for a vax in a immunization CarePlan. @dubdabasoduba @f-odhiambo are these the same thing in our model?

I haven't thought about it that hard, but I don't think these can be the same thing unless we're pre-generating all the tasks. Because along the lines of what you said, we usually have views in the app that show all the vaxs for the patient, and if we're using tasks to track that we'd need to pre-gen to see that correctly. But if we're not using tasks, how is that schedule being encoded in the CarePlan?
